File: Server/src/main/content/global/skill/production/runecrafting/Talisman.java
package content.global.skill.production.runecrafting;

import core.game.node.entity.player.Player;
import core.game.node.item.Item;
import core.game.world.map.Location;

/**
 * The enum Talisman.
 */
public enum Talisman {
    /**
     * The Air.
     */
    AIR(new Item(1438), MysteriousRuin.AIR),
    /**
     * The Mind.
     */
    MIND(new Item(1448), MysteriousRuin.MIND),
    /**
     * The Water.
     */
    WATER(new Item(1444), MysteriousRuin.WATER),
    /**
     * The Earth.
     */
    EARTH(new Item(1440), MysteriousRuin.EARTH),
    /**
     * The Fire.
     */
    FIRE(new Item(1442), MysteriousRuin.FIRE),
    /**
     * The Elemental.
     */
    ELEMENTAL(new Item(5516), null),
    /**
     * The Body.
     */
    BODY(new Item(1446), MysteriousRuin.BODY),
    /**
     * The Cosmic.
     */
    COSMIC(new Item(1454), MysteriousRuin.COSMIC),
    /**
     * The Chaos.
     */
    CHAOS(new Item(1452), MysteriousRuin.CHAOS),
    /**
     * The Nature.
     */
    NATURE(new Item(1462), MysteriousRuin.NATURE),
    /**
     * The Law.
     */
    LAW(new Item(1458), MysteriousRuin.LAW),
    /**
     * The Death.
     */
    DEATH(new Item(1456), MysteriousRuin.DEATH),
    /**
     * The Blood.
     */
    BLOOD(new Item(1450), null),
    /**
     * The Soul.
     */
    SOUL(new Item(1460), null);

    Talisman(final Item talisman, MysteriousRuin ruin) {
        this.talisman = talisman;
        this.ruin = ruin;
    }

    private final Item talisman;
    private final MysteriousRuin ruin;

    /**
     * Gets talisman.
     *
     * @return the talisman
     */
    public Item getTalisman() {
        return talisman;
    }

    /**
     * Locate.
     *
     * @param player the player
     */
    public final void locate(final Player player) {
        if (this == ELEMENTAL || getRuin() == null) {
            player.getPacketDispatch().sendMessage("You cannot tell which direction the Talisman is pulling...");
            return;
        }
        String direction = "";
        Location loc = getRuin().getBase();
        if (player.getLocation().getY() > loc.getY() && player.getLocation().getX() - 1 > loc.getX())
            direction = "south-west";
        else if (player.getLocation().getX() < loc.getX() && player.getLocation().getY() > loc.getY())
            direction = "south-east";
        else if (player.getLocation().getX() > loc.getX() + 1 && player.getLocation().getY() < loc.getY())
            direction = "north-west";
        else if (player.getLocation().getX() < loc.getX() && player.getLocation().getY() < loc.getY())
            direction = "north-east";
        else if (player.getLocation().getY() < loc.getY())
            direction = "north";
        else if (player.getLocation().getY() > loc.getY())
            direction = "south";
        else if (player.getLocation().getX() < loc.getX() + 1)
            direction = "east";
        else if (player.getLocation().getX() > loc.getX() + 1)
            direction = "west";
        player.getPacketDispatch().sendMessage("The talisman pulls towards the " + direction + ".");
    }

    /**
     * Gets ruin.
     *
     * @return the ruin
     */
    public MysteriousRuin getRuin() {
        for (MysteriousRuin ruin : MysteriousRuin.values()) {
            if (ruin.name().equals(name())) {
                return ruin;
            }
        }
        return ruin;
    }

    /**
     * Gets tiara.
     *
     * @return the tiara
     */
    public Tiara getTiara() {
        for (Tiara tiara : Tiara.values()) {
            if (tiara.name().equals(name())) {
                return tiara;
            }
        }
        return null;
    }

    /**
     * For item talisman.
     *
     * @param item the item
     * @return the talisman
     */
    public static Talisman forItem(final Item item) {
        for (Talisman talisman : values()) {
            if (talisman.getTalisman().getId() == item.getId()) {
                return talisman;
            }
        }
        return null;
    }

    /**
     * For name talisman.
     *
     * @param name the name
     * @return the talisman
     */
    public static Talisman forName(final String name) {
        for (Talisman talisman : values()) {
            if (talisman.name().equals(name)) {
                return talisman;
            }
        }
        return null;
    }

}

File: Emulator/src/main/content/global/skill/production/runecrafting/data/Talisman.kt
package content.global.skill.production.runecrafting.data

import org.rs.consts.Items
import core.game.node.entity.player.Player
import core.game.node.item.Item

/**
 * Represents a talisman.
 */
enum class Talisman(val talisman: Item, private val ruin: MysteriousRuin?) {
    AIR(
        talisman = Item(Items.AIR_TALISMAN_1438),
        ruin = MysteriousRuin.AIR
    ),
    MIND(
        talisman = Item(Items.MIND_TALISMAN_1448),
        ruin = MysteriousRuin.MIND
    ),
    WATER(
        talisman = Item(Items.WATER_TALISMAN_1444),
        ruin = MysteriousRuin.WATER
    ),
    EARTH(
        talisman = Item(Items.EARTH_TALISMAN_1440),
        ruin = MysteriousRuin.EARTH
    ),
    FIRE(
        talisman = Item(Items.FIRE_TALISMAN_1442),
        ruin = MysteriousRuin.FIRE
    ),
    ELEMENTAL(
        talisman = Item(Items.ELEMENTAL_TALISMAN_5516),
        ruin = null
    ),
    BODY(
        talisman = Item(Items.BODY_TALISMAN_1446),
        ruin = MysteriousRuin.BODY
    ),
    COSMIC(
        talisman = Item(Items.COSMIC_TALISMAN_1454),
        ruin = MysteriousRuin.COSMIC
    ),
    CHAOS(
        talisman = Item(Items.CHAOS_TALISMAN_1452),
        ruin = MysteriousRuin.CHAOS
    ),
    NATURE(
        talisman = Item(Items.NATURE_TALISMAN_1462),
        ruin = MysteriousRuin.NATURE
    ),
    LAW(
        talisman = Item(Items.LAW_TALISMAN_1458),
        ruin = MysteriousRuin.LAW
    ),
    DEATH(
        talisman = Item(Items.DEATH_TALISMAN_1456),
        ruin = MysteriousRuin.DEATH
    ),
    BLOOD(
        talisman = Item(Items.BLOOD_TALISMAN_1450),
        ruin = MysteriousRuin.BLOOD
    ),
    SOUL(
        talisman = Item(Items.SOUL_TALISMAN_1460),
        ruin = null
    );

    /**
     * Method used to locate a ruin.
     */
    fun locate(player: Player) {
        if (this == ELEMENTAL || ruin == null) {
            player.packetDispatch.sendMessage("You cannot tell which direction the Talisman is pulling...")
            return
        }
        val loc = ruin.base
        val direction = when {
            player.location.y > loc.y && player.location.x - 1 > loc.x -> "south-west"
            player.location.x < loc.x && player.location.y > loc.y -> "south-east"
            player.location.x > loc.x + 1 && player.location.y < loc.y -> "north-west"
            player.location.x < loc.x && player.location.y < loc.y -> "north-east"
            player.location.y < loc.y -> "north"
            player.location.y > loc.y -> "south"
            player.location.x < loc.x + 1 -> "east"
            player.location.x > loc.x + 1 -> "west"
            else -> "unknown direction"
        }
        player.packetDispatch.sendMessage("The talisman pulls towards the $direction.")
    }

    /**
     * Gets the ruin.
     *
     * @return The ruin.
     */
    fun getRuin(): MysteriousRuin? {
        return ruin
    }

    /**
     * Gets the tiara.
     *
     * @return the tiara.
     */
    val tiara: Tiara?
        get() = Tiara.values().find { it.name == name }

    companion object {
        /**
         * Method used to get the [Talisman] by the item.
         *
         * @param item the item.
         * @return the [Talisman] or `null`.
         */
        @JvmStatic
        fun forItem(item: Item): Talisman? {
            return values().find { it.talisman.id == item.id }
        }

        /**
         * Method used to get the [Talisman] by the name.
         *
         * @return the [Talisman] or `null`.
         */
        @JvmStatic
        fun forName(name: String): Talisman? {
            return values().find { it.name == name }
        }
    }
}
